server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationFailuresResetService.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.routing.allocation.command.AllocationCommands;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+
+/**
+ * A service that resets allocation failures counter and triggers reroute, when certain conditions change.
+ * For example cluster version changed or new nodes joined cluster.
+ */
+public record AllocationFailuresResetService(ClusterService clusterService, AllocationService allocationService)
+    implements
+        ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(AllocationFailuresResetService.class);
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.state().getRoutingNodes().hasAllocationFailures()) {
+            if (event.oldestIndexVersionChanged() || event.nodesAdded()) {
+                submitUnbatchedStateUpdateTask(new ClusterStateUpdateTask() {
+                    @Override
+                    public ClusterState execute(ClusterState currentState) {
+                        return allocationService.reroute(currentState, new AllocationCommands(), false, true, false, ActionListener.noop())
+                            .clusterState();
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.warn(() -> "cluster update failure: reset allocation failures counter", e);
+                    }
+                });
+            }
+        }
+    }
+
+    @SuppressForbidden(reason = "legacy usage of unbatched task")
+    private void submitUnbatchedStateUpdateTask(ClusterStateUpdateTask task) {
+        clusterService.submitUnbatchedStateUpdateTask("reset-allocation-failures", task);
+    }
```

Comment:
We would like to remove unbatched state update tasks eventually, lets avoid new usages.

To make a task batch-able you will need:
* A task queue: https://github.com/elastic/elasticsearch/blob/0a410b52161d060d69444d7253026db3a391bea2/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java#L161-L165
* A class/record that captures all the data required to compute the next cluster state: https://github.com/elastic/elasticsearch/blob/0a410b52161d060d69444d7253026db3a391bea2/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java#L327-L343
* An executor class that can process multiple task classes/records at once and apply them on a cluster state (here it will be enough to process a single tasks as they all will be identical?): https://github.com/elastic/elasticsearch/blob/0a410b52161d060d69444d7253026db3a391bea2/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java#L345-L353
* And finally submit the task: https://github.com/elastic/elasticsearch/blob/0a410b52161d060d69444d7253026db3a391bea2/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java#L253-L255

Reply:
I talked with David, he believes using un-batched version is ok here.

Reply:
Hmm now that I look at the implementation, I think we should consider batching here. Otherwise we could end up adding a sequence of nodes and then fruitlessly rerouting N times when once would be enough.

Reply:
Does it make sense to use RerouteService? But it will require changes there to support counter reset flag

Reply:
I think calling `AllocationService#reroute` ourselves is fine here.